core/tx_pool.go

```diff
@@ -1161,84 +1169,94 @@ func (pool *TxPool) scheduleReorgLoop() {
-	pool.truncatePending()
-	pool.truncateQueue()
 
-	// Update all accounts to the latest known pending nonce
-	for addr, list := range pool.pending {
-		highestPending := list.LastElement()
-		pool.pendingNonces.set(addr, highestPending.Nonce()+1)
-	}
-	pool.mu.Unlock()
+			var promoteAddrs []common.InternalAddress
+			if dirtyAccounts != nil && reset == nil {
+				// Only dirty accounts need to be promoted, unless we're resetting.
+				// For resets, all addresses in the tx queue will be promoted and
+				// the flatten operation can be avoided.
+				promoteAddrs = dirtyAccounts.flatten()
+			}
+			pool.mu.Lock()
+			if reset != nil {
+				// Reset from the old head to the new, rescheduling any reorged transactions
+				pool.reset(reset.oldHead, reset.newHead)
+
+				// Nonces were reset, discard any events that became stale
+				for addr := range events {
+					events[addr].Forward(pool.pendingNonces.get(addr))
+					if events[addr].Len() == 0 {
+						delete(events, addr)
+					}
+				}
+				// Reset needs promote for all addresses
+				promoteAddrs = make([]common.InternalAddress, 0, len(pool.queue))
+				for addr := range pool.queue {
+					promoteAddrs = append(promoteAddrs, addr)
+				}
+			}
+			// Check for pending transactions for every account that sent new ones
+			promoted := pool.promoteExecutables(promoteAddrs)
+
+			// If a new block appeared, validate the pool of pending transactions. This will
+			// remove any transaction that has been included in the block or was invalidated
+			// because of another transaction (e.g. higher gas price).
+			if reset != nil {
+				pool.demoteUnexecutables()
+				if reset.newHead != nil {
+					pendingBaseFee := misc.CalcBaseFee(pool.chainconfig, reset.newHead)
+					pool.priced.SetBaseFee(pendingBaseFee)
+				}
+			}
+			// Ensure pool.queue and pool.pending sizes stay within the configured limits.
+			pool.truncatePending()
+			pool.truncateQueue()
+
+			// Update all accounts to the latest known pending nonce
+			for addr, list := range pool.pending {
+				highestPending := list.LastElement()
+				pool.pendingNonces.set(addr, highestPending.Nonce()+1)
+			}
+			pool.mu.Unlock()
 
-	// Notify subsystems for newly added transactions
-	for _, tx := range promoted {
-		addr, _ := types.Sender(pool.signer, tx)
-		internal, err := addr.InternalAddress()
-		if err != nil {
-			log.Debug("Failed to add transaction event", "err", err)
-			continue
-		}
-		if _, ok := events[internal]; !ok {
-			events[(internal)] = newTxSortedMap()
-		}
-		events[internal].Put(tx)
-	}
-	if len(events) > 0 {
-		var txs []*types.Transaction
-		for _, set := range events {
-			txs = append(txs, set.Flatten()...)
+			// Notify subsystems for newly added transactions
+			for _, tx := range promoted {
+				addr, _ := types.Sender(pool.signer, tx)
+				internal, err := addr.InternalAddress()
+				if err != nil {
+					log.Debug("Failed to add transaction event", "err", err)
```

Comment:
Is debug the right level for this? Does it happen very often?

Reply:
Happens everytime someone tries to send an external address